File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:

    int trap(vector<int>& height) {
        int n = height.size();
        int waterTrapped = 0;
        int maxLeft = 0;
        int maxRight = 0;
        int l = 0;
        int r = n-1;
        
        while(l <= r){
            if(height[l] <= height[r]){
                if(height[l] >= maxLeft) maxLeft = height[l];
                else waterTrapped += (maxLeft - height[l]);
                
                l++;
            }else{
                if(height[r] >= maxRight) maxRight = height[r];
                else waterTrapped += (maxRight - height[r]);
                
                r--;
            }
        }
        
        
        return waterTrapped;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        maxL, maxR = height[0], height[-1]
        waterTrapped = 0
        
        left, right = 0, len(height)-1
        
        while left<right:
            if maxL <= maxR:
                left += 1
                maxL = max(maxL, height[left])
                waterTrapped += (maxL-height[left])
            else:
                right -= 1
                maxR = max(maxR, height[right])
                waterTrapped += (maxR-height[right])
        
        return waterTrapped